drivers/xen/regions.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/xen/generic.h>
+
+#include <kernel_arch_interface.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/arch_interface.h>
+#include <zephyr/sys/mem_blocks.h>
+
+LOG_MODULE_REGISTER(xen_regions);
+
+#define EXTENDED_REGIONS_IDX 1
+#define XEN_HYP_NODE DT_INST(0, xen_xen)
+
+static K_MUTEX_DEFINE(ext_regions_lock);
```

Comment:
if it is a common mechanism for allocating memory it is really easy to understand that it can happen even from IRQ context
I would appreciate if you could change it to sync mechanism that we can use not only in threads

Reply:
I wouldn't call it "common mechanism". It has specific usage for the foreign pages of XEN. As for me - I would have removed this mutex at all and leave a comment that allocations between buffers should be syncronized externally. As for me operation is too heavy for spinlock. What do you think?

Reply:
ok, than leave the mutex if it isn't commom allocator
moreover mutex lock should assert in case when we call it from atomic context